### PR TITLE
Handles invalid UTF-8 characters in MARC records

### LINKED ISSUE #1677: Traject error processing certain Unicode characters
This issue is a more general version of the problem detected in #1674

When Traject runs into a record with invalid UTF-8 characters it skips the entire batch of records. We should update it to (1) skip only the record that causes problems and (2) log the error in a way that we can easily identify it.

The culprit is an invalid character in the `language_code_s` field, the following code shows the error:

```
JSON.generate({"language_code_s"=>"\x81ru"}
*** JSON::GeneratorError Exception: source sequence is illegal/malformed utf-8
```

### PR DESCRIPTION
**Pull request by hectorcorrea**
Fixes #1677 

Adds a new Traject writer (`PulSolrJsonWriter`) that gracefully handles records with invalid UTF-8 characters. Notice that you need to explicitly request this new writer when invoking Traject via the `-w` parameter.

```
bundle exec traject -c marc_to_solr/lib/traject_config.rb $FILE -u $SOLR_URL -w Traject::PulSolrJsonWriter
```

Processing a file with the invalid record will show an output similar to this:

```
2021-09-29T10:47:40-04:00  INFO Reading from /Users/correah/Downloads/incremental_15001905180006421_20210923_020921[050]_new
2021-09-29T10:47:40-04:00  INFO    Traject::PulSolrJsonWriter writing to 'http://localhost:8983/solr/puldata/update/json' in batches of 100 with 1 bg threads
2021-09-29T10:47:40-04:00  INFO    Indexer with 1 processing threads, reader: AlmaReader and writer: Traject::PulSolrJsonWriter
2021-09-29T10:47:41-04:00  INFO Seeding the cache for figgy.princeton.edu using Solr...
2021-09-29T10:47:44-04:00 ERROR JSON error in Solr batch add. Will retry documents individually at performance penalty: source sequence is illegal/malformed utf-8
2021-09-29T10:47:46-04:00 ERROR Could not add record 99125394551506421/99125394551506421: source sequence is illegal/malformed utf-8
2021-09-29T10:47:47-04:00  INFO Record 99124679253506421 would be deleted
...
2021-09-29T10:47:47-04:00  INFO Record 99125387689706421 would be deleted
2021-09-29T10:47:47-04:00  INFO finished Indexer#process: 953 records in 7.232 seconds; 131.8 records/second overall.
```

Notice the two ERROR messages logged.

### CODE REVIEW

File: marc_to_solr/lib/pul_solr_json_writer.rb [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+      @skipped_record_incrementer.increment
+      if @max_skipped and skipped_record_count > @max_skipped
+        raise RuntimeError.new("#{self.class.name}: Exceeded maximum number of skipped records (#{@max_skipped}): aborting")
+      end
+
+    end
+  # This rescue clause is not in the original Traject::SolrJsonWriter
+  # Special handling for single record with Unicode characters that JSON.parse() cannot handle.
+  rescue JSON::GeneratorError => ex
+    logger.error "Could not add record #{c.source_record_id}: #{ex}"
```
Comment (hectorcorrea) [inline marc_to_solr/lib/pul_solr_json_writer.rb:177]: Code that I added.

File: marc_to_solr/lib/pul_solr_json_writer.rb [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+require 'uri'
+require 'thread'     # for Mutex/Queue
+require 'concurrent' # for atomic_fixnum
+
+# ==========================
+# This class is a copy of Traject's SolrJsonWriter
+# (https://github.com/traject/traject/blob/master/lib/traject/solr_json_writer.rb)
+# but it has been updated to skip records with Unicode values that Ruby cannot
+# handle in JSON.parse() - see https://github.com/pulibrary/bibdata/issues/1677
+# ==========================
```
Comment (hectorcorrea) [inline marc_to_solr/lib/pul_solr_json_writer.rb:19]: README

File: marc_to_solr/lib/pul_solr_json_writer.rb [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+require 'uri'
+require 'thread'     # for Mutex/Queue
+require 'concurrent' # for atomic_fixnum
+
+# ==========================
+# This class is a copy of Traject's SolrJsonWriter
+# (https://github.com/traject/traject/blob/master/lib/traject/solr_json_writer.rb)
+# but it has been updated to skip records with UTF-8 values that Ruby cannot
+# handle in JSON.parse() - see https://github.com/pulibrary/bibdata/issues/1677
+# ==========================
```
Comment (hectorcorrea) [inline marc_to_solr/lib/pul_solr_json_writer.rb:19]: README first

File: spec/fixtures/files/alma/full_dump/three_records_one_bad_utf8.xml [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+   <subfield code="z">pa</subfield>
+   <subfield code="d">2021-09-23 00:52:11</subfield>
+   <subfield code="p">32101108804400</subfield>
+   <subfield code="y">recap</subfield>
+  </datafield>
+ </record>
+ <record>
+  <leader>01735nam a2200361 i 4500</leader>
+  <controlfield tag="005">20210421120000.0</controlfield>
+  <controlfield tag="008">210421s2021    ru ac    b    001 0сrus d</controlfield>
```
Comment (hectorcorrea) [inline spec/fixtures/files/alma/full_dump/three_records_one_bad_utf8.xml:181]: The `c` in `0crus` is in fact an invalid UTF-8 character (it is represented as \x81 in Ruby) that causes Ruby to throw an exception when attempting to JSON.parse() a string that contains this value.

**Review by christinach — APPROVED** [review]
we indexed and tested it on staging and it looks good.

**Pull request merged**